aptos-move/framework/src/natives/cryptography/algebra/arithmetics/scalar_mul.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright © Aptos Foundation
+
+use crate::{
+    abort_unless_feature_flag_enabled,
+    natives::{
+        cryptography::algebra::{
+            abort_invariant_violated, gas::GasParameters, AlgebraContext, Structure,
+            MOVE_ABORT_CODE_INPUT_VECTOR_SIZES_NOT_MATCHING, MOVE_ABORT_CODE_NOT_IMPLEMENTED,
+        },
+        helpers::{SafeNativeContext, SafeNativeError, SafeNativeResult},
+    },
+    safe_borrow_element, safely_pop_arg, store_element, structure_from_ty_arg,
+};
+use aptos_types::on_chain_config::FeatureFlag;
+use ark_ec::{CurveGroup, Group};
+use ark_ff::Field;
+use move_core_types::gas_algebra::NumArgs;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Value};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, rc::Rc};
+
+fn feature_flag_of_group_scalar_mul(
+    group_opt: Option<Structure>,
+    scalar_field_opt: Option<Structure>,
+) -> Option<FeatureFlag> {
+    match (group_opt, scalar_field_opt) {
+        (Some(Structure::BLS12381G1Affine), Some(Structure::BLS12381Fr))
+        | (Some(Structure::BLS12381G2Affine), Some(Structure::BLS12381Fr))
+        | (Some(Structure::BLS12381Gt), Some(Structure::BLS12381Fr)) => {
+            Some(FeatureFlag::BLS12_381_STRUCTURES)
+        },
+        _ => None,
+    }
+}
+
+macro_rules! abort_unless_group_scalar_mul_enabled {
+    ($context:ident, $group_opt:expr, $scalar_field_opt:expr) => {
+        let flag_opt = feature_flag_of_group_scalar_mul($group_opt, $scalar_field_opt);
+        abort_unless_feature_flag_enabled!($context, flag_opt);
+    };
+}
+
+macro_rules! ark_scalar_mul_internal {
```

Comment:
Shouldn't this be in `aptos-move/framework/src/natives/cryptography/algebra/arithmetics/mod.rs` with everything else?